xarray/testing/strategies.py

```diff
@@ -0,0 +1,452 @@
+from collections.abc import Hashable, Iterable, Mapping, Sequence
+from typing import TYPE_CHECKING, Any, Protocol, Union, overload
+
+import hypothesis.extra.numpy as npst
+import hypothesis.strategies as st
+import numpy as np
+from hypothesis.errors import InvalidArgument
+
+import xarray as xr
+from xarray.core.types import T_DuckArray
+
+if TYPE_CHECKING:
+    from xarray.core.types import _DTypeLikeNested, _ShapeLike
+
+
+__all__ = [
+    "supported_dtypes",
+    "names",
+    "dimension_names",
+    "dimension_sizes",
+    "attrs",
+    "variables",
+    "unique_subset_of",
+]
+
+
+class ArrayStrategyFn(Protocol[T_DuckArray]):
+    def __call__(
+        self,
+        *,
+        shape: "_ShapeLike",
+        dtype: "_DTypeLikeNested",
+    ) -> st.SearchStrategy[T_DuckArray]:
+        ...
+
+
+def supported_dtypes() -> st.SearchStrategy[np.dtype]:
+    """
+    Generates only those numpy dtypes which xarray can handle.
+
+    Avoiding using hypothesis.extra.numpy.scalar_dtypes is required to exclude weirder dtypes e.g. unicode, byte_string, array, or nested dtypes.
+    Also required to dodge bugs with pandas non-nanosecond datetime overflows.
+    Note does not generate all dtypes that xarray can handle - just only generates dtypes which it definitely can.
+
+    Requires the hypothesis package to be installed.
+    """
+    # TODO should this be exposed publicly?
+    # We should at least decide what the set of numpy dtypes that xarray officially supports is.
+    return (
+        npst.integer_dtypes()
+        | npst.unsigned_integer_dtypes()
+        | npst.floating_dtypes()
+        | npst.complex_number_dtypes()
+    )
```

Comment:
we do support string dtypes, but only for a subset of operations. Is this worth mentioning?

Reply:
This is not meant to be an exhaustive list (yet). It doesn't include datetimes either.

Reply:
agreed, but most operations don't make sense on string or datetime dtypes so it might be better to make a separate list of dtypes for those?

Reply:
Sure - I'm just saying let's defer detailed discussions of which types to test until another issue / PR, the point of this PR is to provide a framework flexible enough to easily test xarray functions with any type we want, which this achieves.